neqo-client/src/main.rs

```diff
@@ -612,21 +734,105 @@ impl<'a> Handler<'a> {
 
         Ok(true)
     }
+
+    fn print_stats(&self) {
+        for (stream_id, (start_time, end_time)) in &self.upload_stats {
+            match (start_time, end_time) {
+                (Some(start), Some(end)) => {
+                    let upload_time = end.duration_since(*start);
+                    println!("Stream ID: {:?}, Upload time: {:?}", stream_id, upload_time);
+                }
+                _ => {
+                    println!("Stream ID: {:?}, Upload failed", stream_id);
+                }
+            }
+        }
+    }
 }
 
-fn to_headers(values: &[impl AsRef<str>]) -> Vec<Header> {
-    values
-        .iter()
-        .scan(None, |state, value| {
-            if let Some(name) = state.take() {
-                *state = None;
-                Some(Header::new(name, value.as_ref()))
-            } else {
-                *state = Some(value.as_ref().to_string());
-                None
+impl<'a> UploadHandler<'a> {
+    fn upload_urls(&mut self, client: &mut Http3Client) {
+        loop {
+            if self.url_queue.is_empty() {
+                break;
             }
-        })
-        .collect()
+            if self.streams.len() >= self.args.concurrency {
+                break;
+            }
+            if !self.upload_next(client) {
+                break;
+            }
+        }
+    }
+
+    fn upload_next(&mut self, client: &mut Http3Client) -> bool {
```

Comment:
There seems to be a lot of overlap between this and the other code.  I'd suggest a refactoring, but it is a little hard to disentangle.  Can you try to find some more commonality?

Reply:
Yeah, I'll try.